Cython/Utility/ExtensionTypes.c

```diff
@@ -681,7 +681,46 @@ static int __Pyx_call_type_traverse(PyObject *o, int always_call, visitproc visi
 }
 #endif
 
+////////////////// LimitedApiGetTypeDict.proto //////////////////////
+
+#if CYTHON_COMPILING_IN_LIMITED_API
+// This is a little hacky - the Limited API works quite hard to stop us getting
+// the dict of a type object. But apparently not not hard enough...
+//
+// In future we should prefer to work with mutable types, and then make them immutable
+// once we're done (pending C API support for this).
+static PyObject *__Pyx_GetTypeDict(PyTypeObject *tp); /* proto */
+#endif
+
+////////////////// LimitedApiGetTypeDict //////////////////////
+
+#if CYTHON_COMPILING_IN_LIMITED_API
+static Py_ssize_t __Pyx_GetTypeDictOffset(void) {
+    // TODO - if we ever support custom metatypes for extension types then
+    // we have to drop this caching.
+    static Py_ssize_t tp_dictoffset = 0;
+    if (tp_dictoffset == 0) {
+        PyObject *tp_dictoffset_o = NULL;
+        tp_dictoffset_o = PyObject_GetAttrString((PyObject*)(&PyType_Type), "__dictoffset__");
+        if (unlikely(!tp_dictoffset_o)) return -1;
+        tp_dictoffset = PyLong_AsSsize_t(tp_dictoffset_o);
+        if (unlikely(tp_dictoffset == -1 && PyErr_Occurred())) {
+            tp_dictoffset = 0; // try again next time
+            return -1;
+        }
+    }
+    return tp_dictoffset;
+}
+
+static PyObject *__Pyx_GetTypeDict(PyTypeObject *tp) {
+    Py_ssize_t tp_dictoffset = __Pyx_GetTypeDictOffset();
+    if (unlikely(tp_dictoffset == -1 && PyErr_Occurred())) return NULL;
```

Comment:
`type.__dictoffset__` is not writable, so it seems a fatal error if `tp_dictoffset < 0` and no exception was raised. That should never happen.
```suggestion
    if (unlikely(tp_dictoffset == -1)) {
        assert(PyErr_Occurred());
        return NULL;
    }
```

Reply:
tp_dictoffset can definitely be negative (and I think always is for `Py_TPFLAGS_MANAGED_DICT`) - the negative value doesn't have a special meaning except that the dict pointer is infront of the object.

On Python 3.11:

```
>>> class C:
...   pass
... 
>>> C.__dictoffset__
-48
>>> 
```

I suspect `-1` is very unlikely because it implies 1 byte pointers. But it isn't obvious that it's always forbidden so I'd rather leave this bit as is.

Reply:
This doesn't look to be completely right... negative offset indicates the offset is from the end of the object instead of the start.